src/serde_arrow_variable_list_array.erl

```diff
@@ -0,0 +1,80 @@
+
+-behaviour(serde_arrow_array).
+
+-export([new/2]).
+
+-include("serde_arrow_array.hrl").
+-spec new(Values :: list(), Type :: map() | serde_arrow_type:arrow_type()) -> Array :: #array{}.
+new(Values, Opts) when is_map(Opts) ->
+    case maps:get(type, Opts, undefined) of
+        undefined ->
+            erlang:error(badarg);
+        Type when is_tuple(Type) orelse is_atom(Type) ->
+            new(Values, Type)
+    end;
+new(Values, GivenType) when
+    (is_tuple(GivenType) andalso tuple_size(GivenType) =:= 2) orelse is_atom(GivenType)
+->
+    Type = serde_arrow_type:normalize(GivenType),
+    %% Data
+    Flattened = serde_arrow_utils:flatten(Values),
+    Array = serde_arrow_fixed_primitive_array:new(Flattened, Type),
+    %% Offsets
+    [0 | FlatOffsets] = serde_arrow_offsets:new_list(Flattened, Type),
+    Offset = serde_arrow_buffer:new([0 | offsets(Values, FlatOffsets, 0)], {s, 32}),
+    init(Values, Type, Array, Offset);
+new(Values, {variable_list, NestedType, undefined} = Type) ->
+    Flattened = serde_arrow_utils:flatten(Values),
+    Array = new(Flattened, NestedType),
+    Offset = serde_arrow_buffer:new(nested_offsets(Values, Type), {s, 32}),
+    init(Values, Type, Array, Offset);
+new(Values, {fixed_list, NestedType, _Size} = Type) ->
+    Flattened = serde_arrow_utils:flatten(Values),
+    Array = serde_arrow_fixed_list_array:new(Flattened, NestedType),
+    Offset = serde_arrow_buffer:new(nested_offsets(Values, Type), {s, 32}),
+    init(Values, Type, Array, Offset).
+
+%%%%%%%%%%%
+%% Utils %%
+%%%%%%%%%%%
+
+-spec init(
+    Values :: list(), Type :: serde_arrow_type:arrow_type(), Data :: #array{}, Offsets :: list()
+) -> #array{}.
+init(Values, Type, Data, Offsets) ->
+    Len = length(Values),
+    {Bitmap, NullCount} = serde_arrow_bitmap:validity_bitmap(Values),
+    #array{
+        layout = variable_list,
+        type = Type,
+        len = Len,
+        element_len = undefined,
+        null_count = NullCount,
+        validity_bitmap = Bitmap,
+        offsets = Offsets,
+        data = Data
+    }.
+
+offsets([H | T], Offsets, CurOffset) when (H =:= undefined) orelse (H =:= nil) ->
+    [CurOffset | offsets(T, Offsets, CurOffset)];
+offsets([H | T], Offsets, _CurOffset) ->
+    Len = length(H),
+    {HOffsets, TOffsets} = lists:split(Len, Offsets),
+    CurOffset = lists:last(HOffsets),
+    [CurOffset | offsets(T, TOffsets, CurOffset)];
+offsets([], _Offset, _CurOffset) ->
+    [].
+
+nested_offsets(List, Type) ->
+    case serde_arrow_utils:nesting(List) of
+        2 ->
+            [0 | FlatOffsets] = serde_arrow_offsets:new_list(
+                serde_arrow_utils:flatten(List), serde_arrow_type:normalize(Type)
+            ),
+            [0 | offsets(List, FlatOffsets, 0)];
+        _Nesting ->
+            Flattened = serde_arrow_utils:flatten(List),
+            [0 | FlatOffsets] = nested_offsets(Flattened, element(2, Type)),
+            [0 | offsets(List, FlatOffsets, 0)]
+    end.
```

Comment:
This function is a disaster. We're computing offsets we have already computed for each nesting. Actually the offsets are already in the metadata of nested array, but they are in the form of a buffer, which stores the values of the binary.

This led to me feel that serializing at the intermediate stage is premature, as the data in the intermediate stage is useful for computing other data. Serializing here would cause `from_erlang` followed by `to_erlang` to take twice as long as `from_erlang` followed by `to_arrow` because we are serializing and then deserializing again. On the other hand not serializing would result in `from_erlang` and `to_arrow` to take the same time (only 1 serialization in each case) and `from_erlang` and `to_erlang` would be orders of magnitude faster as there in no serialization or deserialization involved.

Now the question is if I should make this change to buffers (in another PR) and rewrite the offset calculation here, or if I should keep buffers as it is and then write a `new_with_offset` function to return a tuple with the offsets and array so we avoid this re-computation.

Reply:
I would favor performance, so I think a new function that does it once, even if it has to reimplement aspects of fixed_list/variable_binary would be best.

Reply:
Both approaches will have the same performance. Approach 1 may lead to perf improvements elsewhere (by making the precomputed stuff accessible) and while Approach 2 is just code duplication to save to precomputed stuff and return it as a 2 tuple.

Going with Approach 1.